### PR TITLE
Moved from stl::set and stl::map to hash tables

### PR DESCRIPTION
**Pull request by arkpar**

### CODE REVIEW

File: libdevcore/Common.h [hunk 1]
```diff
@@ -215,4 +222,14 @@ inline dev::WithExisting max(dev::WithExisting _a, dev::WithExisting _b)
 	return static_cast<dev::WithExisting>(max(static_cast<int>(_a), static_cast<int>(_b)));
 }
 
+template <> struct hash<dev::u256>
+{
+	size_t operator()(dev::u256 const& _a) const
+	{
+		unsigned size = _a.backend().size();
+		auto limbs = _a.backend().limbs();
+		return boost::hash_range(limbs, limbs + size);
```
Comment (chriseth) [inline libdevcore/Common.h]: Is it ok to use them this way? I.e. two equal numbers always have equal data in their limbs, they never contain "dirty" bytes?

File: libdevcore/FixedHash.h [hunk 2]
```diff
@@ -215,6 +208,23 @@ template<> inline bool FixedHash<32>::operator==(FixedHash<32> const& _other) co
 	return (hash1[0] == hash2[0]) && (hash1[1] == hash2[1]) && (hash1[2] == hash2[2]) && (hash1[3] == hash2[3]);
 }
 
+/// Fast std::hash compatible hash function object for h64.
+template<> inline size_t FixedHash<8>::hash::operator()(FixedHash<8> const& value) const
+{
+	const uint64_t*data = (const uint64_t*)value.data();
```
Comment (chriseth) [inline libdevcore/FixedHash.h]: Spacing and position of the `const` - also in various other places.

File: libdevcore/FixedHash.h [hunk 3]
```diff
@@ -215,6 +208,23 @@ template<> inline bool FixedHash<32>::operator==(FixedHash<32> const& _other) co
 	return (hash1[0] == hash2[0]) && (hash1[1] == hash2[1]) && (hash1[2] == hash2[2]) && (hash1[3] == hash2[3]);
 }
 
+/// Fast std::hash compatible hash function object for h64.
+template<> inline size_t FixedHash<8>::hash::operator()(FixedHash<8> const& value) const
```
Comment (chriseth) [inline libdevcore/FixedHash.h]: The previous implementations mangled the values, the new implementation does not.
Can we assume that `FixedHash::hash` is never used on "non-hashed" values?

File: libdevcrypto/Common.h [hunk 4]
```diff
@@ -68,8 +68,8 @@ extern Address ZeroAddress;
 /// A vector of Ethereum addresses.
 using Addresses = h160s;
 
-/// A set of Ethereum addresses.
-using AddressSet = std::set<h160>;
+/// A hash table of Ethereum addresses.
```
Comment (chriseth) [inline libdevcrypto/Common.h]: For me, "hash table" is an associative array, i.e. an `unordered_map`.

File: libp2p/Common.h [hunk 5]
```diff
@@ -203,3 +203,26 @@ struct Node
...
+	{
+		if (_a.is_v4())
+			return _a.to_v4().to_ulong();
+		if (_a.is_v6())
+		{
+			auto const& range = _a.to_v6().to_bytes();
+			return boost::hash_range(range.begin(), range.end());
+		}
+		if (_a.is_unspecified())
+			return static_cast<size_t>(0x3487194039229152ul);  // Some random value
```
Comment (chriseth) [inline libp2p/Common.h]: "chosen by fair dice roll, guaranteed to be random" scnr ;-)

File: libp2p/Common.h [hunk 6]
```diff
@@ -203,3 +203,26 @@ struct Node
...
+
+/// std::hash for asio::adress
+namespace std
+{
+
+template <> struct hash<bi::address>
+{
+	size_t operator()(bi::address const& _a) const
+	{
+		if (_a.is_v4())
```
Comment (chriseth) [inline libp2p/Common.h]: You mangle the data for v6 but not for v4. What about using the built-in hash function for v4?

File: libp2p/NodeTable.cpp [hunk 7]
```diff
@@ -293,10 +293,9 @@ vector<shared_ptr<NodeEntry>> NodeTable::nearestNodeEntries(NodeId _target)
 		}
 	
 	vector<shared_ptr<NodeEntry>> ret;
-	for (auto& nodes: found)
-		for (auto n: nodes.second)
-			if (n->endpoint.isAllowed())
-				ret.push_back(n);
+		for (auto n: found)
```
Comment (chriseth) [inline libp2p/NodeTable.cpp]: Indentation changed.

File: libdevcore/Common.h (hunk 1 above)
Comment (arkpar) [inline libdevcore/Common.h]: Current boost code guarantees it. I.e. same functions are used internally for comparison. This is implementation detail though and may change in future versions. For now this is the fastest way

File: libp2p/Common.h (hunk 6 above)
Comment (arkpar) [inline libp2p/Common.h]: Fixed

**Comment by arkpar** [comment]
@subtly please take a look at libp2p changes

**Comment by chriseth** [comment]
looksgood from my side

**Comment by subtly** [comment]
The network tests aren't run by built bot. Do they all pass when run locally (on mac/linux/win)?

File: libp2p/Common.h [hunk 8]
```diff
@@ -214,3 +214,26 @@ struct Node
...
+			return std::hash<unsigned long>()(_a.to_v4().to_ulong());
+		if (_a.is_v6())
+		{
+			auto const& range = _a.to_v6().to_bytes();
+			return boost::hash_range(range.begin(), range.end());
+		}
+		if (_a.is_unspecified())
+			return static_cast<size_t>(0x3487194039229152ul);  // Chosen by fair dice roll, guaranteed to be random
```
Comment (subtly) [inline libp2p/Common.h]: IS this outside the range that would be returned by IPv4 and IPv6 address?

Comment (arkpar) [inline libp2p/Common.h]: Can collide with a ipv6 address hash, but this is not a problem since unordered_set/map handles collisions

**Comment by arkpar** [comment]
Some network tests fail randomly, although this is not caused by this PR. 
```
⧎ ✘  13:30:55|p2p  Couldn't start accepting connections on host. Failed to accept socket.
 Throw location unknown (consider using BOOST_THROW_EXCEPTION)
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >
std::exception::what: bind: Address already in use

terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >'
  what():  bind: Address already in use
```
Looks like the socket was not released by previous test

**Comment by subtly** [comment]
All network tests pass on macos and linux.

File: libp2p/NodeTable.cpp [hunk 9]
```diff
@@ -280,10 +280,9 @@ vector<shared_ptr<NodeEntry>> NodeTable::nearestNodeEntries(NodeId _target)
 		}
 	
 	vector<shared_ptr<NodeEntry>> ret;
-	for (auto& nodes: found)
-		for (auto n: nodes.second)
-			if (ret.size() < s_bucketSize && !!n->endpoint && n->endpoint.isAllowed())
-				ret.push_back(n);
+	for (auto n: found)
```
Comment (subtly) [inline libp2p/NodeTable.cpp]: This changes the functionality of previous code. The code intentionally populates ret in order of the maps' key index.

File: libp2p/NodeTable.h [hunk 10]
```diff
@@ -243,32 +243,32 @@ class NodeTable: UDPSocketEvents, public std::enable_shared_from_this<NodeTable>
...
+	std::unordered_map<NodeId, std::shared_ptr<NodeEntry>> m_nodes;	///< Nodes
 
-	mutable Mutex x_state;									///< LOCK x_state first if both x_nodes and x_state locks are required.
-	std::array<NodeBucket, s_bins> m_state;					///< State of p2p node network.
+	mutable Mutex x_state;											///< LOCK x_state first if both x_nodes and x_state locks are required.
+	std::array<NodeBucket, s_bins> m_state;							///< State of p2p node network.
 
-	Mutex x_evictions;										///< LOCK x_evictions first if both x_nodes and x_evictions locks are required.
-	std::deque<EvictionTimeout> m_evictions;					///< Eviction timeouts.
+	Mutex x_evictions;												///< LOCK x_nodes first if both x_nodes and x_evictions locks are required.
```
Comment (subtly) [inline libp2p/NodeTable.h]: Rebase is required? This was changed recently.

Comment (arkpar) [inline libp2p/NodeTable.h]: Rebased.

**Pull request merged**